sls_api/endpoints/tools/files.py

```diff
@@ -1075,3 +1136,187 @@ def handle_collection_toc(project, collection_id, language=None):
+
+        # Read file contents and parse json into a dict
+        with open(filepath, "r", encoding="utf-8-sig") as json_file:
+            toc = json.load(json_file)
+
+    except FileNotFoundError:
+        logger.exception(f"File not found error when trying to read ToC-file at {filepath}.")
+        return create_error_response(f"Error: {filename} not found.", 404)
+    except PermissionError:
+        logger.exception(f"Permission denied error when trying to read ToC-file at {filepath}.")
+        return create_error_response(f"Error: permission denied when trying to read {filename}.", 403)
+    except json.JSONDecodeError:
+        logger.exception(f"Invalid JSON in ToC file at {filepath}.")
+        return create_error_response(f"Error: file {filename} contains invalid JSON.", 415)
+    except Exception:
+        logger.exception(f"Error accessing file at {filepath}.")
+        return create_error_response(f"Error accessing file at {filepath}.", 500)
+
+    if not toc:
+        return create_error_response(f"The table of contents file {filename} is empty.")
+
+    # Fetch publication data
+    collection_table = get_table("publication_collection")
+    publication_table = get_table("publication")
+
+    try:
+        with db_engine.connect() as connection:
+            # Check for publication_collection existence in project
+            select_coll_stmt = (
+                select(collection_table.c.id)
+                .where(collection_table.c.id == collection_id)
+                .where(collection_table.c.project_id == project_id)
+                .where(collection_table.c.deleted < 1)
+            )
+            result = connection.execute(select_coll_stmt).first()
+
+            if not result:
+                return create_error_response("Validation error: could not find publication collection, either 'project' or 'collection_id' is invalid.")
+
+            # Proceed to selecting the publications
+            select_pub_stmt = (
+                select(
+                    publication_table.c.id,
+                    publication_table.c.name,
+                    publication_table.c.original_publication_date
+                )
+                .where(publication_table.c.publication_collection_id == collection_id)
+                .where(publication_table.c.deleted < 1)
+            )
+            publication_rows = connection.execute(select_pub_stmt).fetchall()
+
+    except Exception:
+        logger.exception("Exception retrieving publications data for collection ToC update.")
+        return create_error_response("Unexpected error: failed to retrieve publications data.", 500)
+
+    if len(publication_rows) == 0:
+        return create_error_response(f"No publications in collection with ID '{collection_id}'.")
+
+    # Build update map
+    update_map = {}
+    for row in publication_rows:
+        item_id = f"{collection_id}_{row['id']}"
```

Comment:
The `row` objects in `publication_rows` here are not dicts, and thus you cannot access their keys with the `['id']` syntax - by default, SQLalchemy returns `Row` objects, which act like named tuples, not like dicts.
In order to access rows properly, you need to either use `row.id` syntax, use `row._mapping['id']`, or call `row._asdict()` to get a dict.

For documentation on the behavior of `Row` objects, see https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row